doctr/models/recognition/transformer/tensorflow.py

```diff
@@ -3,263 +3,177 @@
+import math
+from typing import Any, Optional, Tuple
 
-
-from typing import Any, Tuple
-
-import numpy as np
 import tensorflow as tf
+from tensorflow.keras import layers
 
-__all__ = ['Decoder', 'positional_encoding', 'create_look_ahead_mask', 'create_padding_mask']
-
-
-def get_angles(pos: np.array, i: np.array, d_model: int = 512) -> np.array:
-    """This function compute the 2D array of angles for sinusoidal positional encoding.
+from doctr.utils.repr import NestedObject
 
-    Args:
-        pos: range of positions to encode
-        i: range of depth to encode positions
-        d_model: depth parameter of the model
+__all__ = ['Decoder', 'PositionalEncoding']
 
-    Returns:
-        2D array of angles, len(pos) x len(i)
-    """
-    angle_rates = 1 / np.power(10000, (2 * (i // 2)) / np.float32(d_model))
-    return pos * angle_rates
+tf.config.run_functions_eagerly(True)
 
 
-def positional_encoding(position: int, d_model: int = 512, dtype=tf.float32) -> tf.Tensor:
-    """This function computes the 2D positional encoding of the position, on a depth d_model
+class PositionalEncoding(layers.Layer, NestedObject):
+    """ Compute positional encoding """
 
-    Args:
-        position: Number of positions to encode
-        d_model: depth of the encoding
+    def __init__(self, d_model: int, dropout: float = 0.1, max_len: int = 5000) -> None:
+        super(PositionalEncoding, self).__init__()
+        self.dropout = layers.Dropout(rate=dropout)
 
-    Returns:
-        2D positional encoding as described in Transformer paper.
-    """
-    angle_rads = get_angles(
-        np.arange(position)[:, np.newaxis],
-        np.arange(d_model)[np.newaxis, :],
-        d_model,
-    )
-    # apply sin to even indices in the array; 2i
-    angle_rads[:, 0::2] = np.sin(angle_rads[:, 0::2])
-    # apply cos to odd indices in the array; 2i+1
-    angle_rads[:, 1::2] = np.cos(angle_rads[:, 1::2])
-    pos_encoding = angle_rads[np.newaxis, ...]
-    return tf.cast(pos_encoding, dtype=dtype)
-
-
-@tf.function
-def create_padding_mask(seq: tf.Tensor, padding: int = 0, dtype=tf.float32) -> tf.Tensor:
-    seq = tf.cast(tf.math.equal(seq, padding), dtype)
-    # add extra dimensions to add the padding to the attention logits.
-    return seq[:, tf.newaxis, tf.newaxis, :]  # (batch_size, 1, 1, seq_len)
+        # Compute the positional encodings once in log space.
+        pe = tf.Variable(tf.zeros((max_len, d_model)))
+        position = tf.cast(
+            tf.expand_dims(tf.experimental.numpy.arange(start=0, stop=max_len), axis=1), dtype=tf.float32
+        )
+        div_term = tf.math.exp(
+            tf.cast(
+                tf.experimental.numpy.arange(start=0, stop=d_model, step=2), dtype=tf.float32
+            ) * -(math.log(10000.0) / d_model)
+        )
+        pe = pe.numpy()
+        pe[:, 0::2] = tf.math.sin(position * div_term)
+        pe[:, 1::2] = tf.math.cos(position * div_term)
```

Comment:
@frgfm Any idea how to do this in TF and remove the numpy cast? :sweat_smile:

Reply:
Hi @felixdittrich92, I know [this](https://www.tensorflow.org/api_docs/python/tf/numpy_function) but I think it is still a bit experimental